pkg/kubelet/winstats/perfcounter_nodestats.go

```diff
@@ -0,0 +1,151 @@
+// +build windows
+
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package winstats
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+	"github.com/lxn/win"
+)
+
+// NewPerfCounterClient creates a client using perf counters
+func NewPerfCounterClient() (Client, error) {
+	return NewClient(&perfCounterNodeStatsClient{})
+}
+
+// perfCounterNodeStatsClient is a client that provides Windows Stats via PerfCounters
+type perfCounterNodeStatsClient struct {
+	nodeStats
+	mu sync.RWMutex
+}
+
+func (p *perfCounterNodeStatsClient) startMonitoring() error {
+	memory, err := getPhysicallyInstalledSystemMemoryBytes()
+	if err != nil {
+		return err
+	}
+	p.nodeStats.memoryPhysicalCapacityBytes = memory
+
+	version, err := exec.Command("cmd", "/C", "ver").Output()
+	if err != nil {
+		return err
+	}
+	p.kernelVersion = strings.TrimSpace(string(version))
+
+	cpuChan, err := readPerformanceCounter(cpuQuery)
+	if err != nil {
+		return err
+	}
+
+	memWorkingSetChan, err := readPerformanceCounter(memoryPrivWorkingSetQuery)
+	if err != nil {
+		return err
+	}
+
+	memCommittedBytesChan, err := readPerformanceCounter(memoryCommittedBytesQuery)
+	if err != nil {
+		return err
+	}
+
+	go p.startNodeMonitoring(cpuChan, memWorkingSetChan, memCommittedBytesChan)
+	return nil
+}
+
+func (p *perfCounterNodeStatsClient) getMachineInfo() (*cadvisorapi.MachineInfo, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	return &cadvisorapi.MachineInfo{
+		NumCores:       runtime.NumCPU(),
+		MemoryCapacity: p.memoryPhysicalCapacityBytes,
```

Comment:
Still accessing the internal `nodeStats` object without acquiring the lock. 
Suggest switch to calling `getNodeStats()` to get a copy.

Reply:
I don't believe a lock is required here because all of the fields accessed from node stats are never written to from another goroutine. If it makes it more clear, I can move to calling `getNodeStats` and use a copy.

Reply:
Technically you are correct, but this makes the code hard to maintain. You have a lock guarding the nodeStats partially (i.e., only some fields), and it's hard to keep track of them. 
If kernel version and capacity are not to change after kubelet starts up, I think it makes sense to define a `nodeInfo` struct for those immutable fields and keep `nodeStats` for the mutable ones. Also add a comment above the lock explaining that it guards the `nodeStats` object.

Reply:
I updated it to include a separate struct `nodeInfo` as you mentioned, keeping separate `nodeMetrics` which needs a lock. Thanks for the suggestion!